targets/TARGET_NUVOTON/TARGET_M2351/trng_api.cpp

```diff
@@ -0,0 +1,157 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017-2018 Nuvoton
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if DEVICE_TRNG
+
+#include "cmsis.h"
+#include <limits.h>
+#include "crypto-misc.h"
+#include "hal/trng_api.h"
+#include "platform/mbed_toolchain.h"
+#include "platform/mbed_critical.h"
+#include "platform/mbed_error.h"
+#include "platform/SingletonPtr.h"
+#include "platform/PlatformMutex.h"
+#include "nu_modutil.h"
+#include "hal_secure.h"
+#include "partition_M2351.h"
+
+#if defined(SCU_INIT_PNSSET5_VAL) && (SCU_INIT_PNSSET5_VAL & (1 << 25))
+#error("We just support secure TRNG")
+#endif
+
+#if defined (__ARM_FEATURE_CMSE) && (__ARM_FEATURE_CMSE == 3L)
+
+/* Module init definition: modname, clkidx, clksrc, clkdiv, rstidx, irqnum, misc */
+static const struct nu_modinit_s trng_modinit = {TRNG_0, TRNG_MODULE, 0, 0, TRNG_RST, TRNG_IRQn, NULL};
+
+/* TRNG init counter. TRNG is kept active as it is non-zero. */
+static uint16_t trng_init_counter = 0U;
+
+/* Mutex for synchronizing access to TRNG H/W */
+static SingletonPtr<PlatformMutex> trng_mutex;
+
+#endif
+
+void trng_init(trng_t *obj)
+{
+    trng_init_s(obj);
+}
+
+void trng_free(trng_t *obj)
+{
+    trng_free_s(obj);
+}
+
+int trng_get_bytes(trng_t *obj, uint8_t *output, size_t length, size_t *output_length)
+{
+    uint32_t output_length_;
+    int32_t rc = trng_get_bytes_s(obj, output, (uint32_t) length, &output_length_);
+    if (output_length) {
+        *output_length = output_length_;
+    }
+    return rc;
+}
+
+#if defined (__ARM_FEATURE_CMSE) && (__ARM_FEATURE_CMSE == 3U)
+
+__NONSECURE_ENTRY
+extern "C"
+void trng_init_s(MBED_UNUSED void *obj)
+{
+    core_util_critical_section_enter();
+    if (trng_init_counter == USHRT_MAX) {
+        core_util_critical_section_exit();
+        MBED_ERROR(MBED_MAKE_ERROR(MBED_MODULE_HAL, MBED_ERROR_CODE_OVERFLOW), \
+                   "TRNG initialization counter would overflow");
+    }
+    core_util_atomic_incr_u16(&trng_init_counter, 1);
+    if (trng_init_counter == 1) {
```

Comment:
Re-reading the counter isn't atomic - you need to look at the return value of the increment function.
```suggestion
    if (core_util_atomic_incr_u16(&trng_init_counter, 1) == 1) {
```

Although here you're inside a critical section anyway, meaning you don't need the atomic increment at all, you can just do `++trng_init_counter`.

Reply:
Fixed